### PR TITLE
DEV: initialize boost submodule in Gitpod Dockerfile

### PR DESCRIPTION
**Pull request by rgommers**
This fixes the build, which is currently broken (see, e.g., https://github.com/scipy/scipy/runs/2588163998).

Also rename the file to be visible (same as in numpy, makes it easier to discover and search).

@trallard I didn't test the Dockerfile, because I'm not sure this is the best way to do it - the whole `COPY` and shallow clone business to keep the image as slim as possible has some subtleties I don't fully understand:) The `git submodule` invocation is optimal though, I did test that. Please check if this is the right way of doing it.

### CODE REVIEW

**Comment by rgommers** [comment]
Grr something else is broken due to a recent commit, adding C++ compile flags somehow messes with the ccache configuration:
```
...
compile options: '-I/opt/conda/envs/scipy-dev/lib/python3.9/site-packages/numpy/core/include -I/opt/conda/envs/scipy-dev/lib/python3.9/site-packages/numpy/core/include -Ibuild/src.linux-x86_64-3.9/numpy/distutils/include -I/opt/conda/envs/scipy-dev/include/python3.9 -c'
extra options: '-msse -msse2 -msse3'
x86_64-conda_cos6-linux-gnu-gcc -pthread -shared -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath,/opt/conda/envs/scipy-dev/lib -Wl,-rpath-link,/opt/conda/envs/scipy-dev/lib -L/opt/conda/envs/scipy-dev/lib -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath,/opt/conda/envs/scipy-dev/lib -Wl,-rpath-link,/opt/conda/envs/scipy-dev/lib -L/opt/conda/envs/scipy-dev/lib -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath,/opt/conda/envs/scipy-dev/lib -Wl,-rpath-link,/opt/conda/envs/scipy-dev/lib -L/opt/conda/envs/scipy-dev/lib -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /opt/conda/envs/scipy-dev/include build/temp.linux-x86_64-3.9/scipy/cluster/_optimal_leaf_ordering.o -Lbuild/temp.linux-x86_64-3.9 -o build/lib.linux-x86_64-3.9/scipy/cluster/_optimal_leaf_ordering.cpython-39-x86_64-linux-gnu.so -Wl,--version-script=build/temp.linux-x86_64-3.9/link-version-scipy.cluster._optimal_leaf_ordering.map
C compiler: ccache /opt/conda/envs/scipy-dev/bin/x86_64-conda-linux-gnu-cc -Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -O2 -Wall -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /opt/conda/envs/scipy-dev/include -fPIC

creating /tmp/tmpzkqouq5o/tmp
creating /tmp/tmpzkqouq5o/tmp/tmpzkqouq5o
compile options: '-I/opt/conda/envs/scipy-dev/include/python3.9 -c'
extra options: '-std=c++14'
ccache: /tmp/tmpzkqouq5o/main.cpp
C compiler: ccache /opt/conda/envs/scipy-dev/bin/x86_64-conda-linux-gnu-cc -Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -O2 -Wall -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /opt/conda/envs/scipy-dev/include -fPIC

creating /tmp/tmp_thsfrac/tmp
creating /tmp/tmp_thsfrac/tmp/tmp_thsfrac
compile options: '-I/opt/conda/envs/scipy-dev/include/python3.9 -c'
ccache: /tmp/tmp_thsfrac/main.c
C compiler: ccache /opt/conda/envs/scipy-dev/bin/x86_64-conda-linux-gnu-cc -Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -O2 -Wall -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /opt/conda/envs/scipy-dev/include -fPIC

creating /tmp/tmp9ya5xnor/tmp
creating /tmp/tmp9ya5xnor/tmp/tmp9ya5xnor
compile options: '-I/opt/conda/envs/scipy-dev/include/python3.9 -c'
extra options: '-pthread'
ccache: /tmp/tmp9ya5xnor/main.c
C compiler: ccache /opt/conda/envs/scipy-dev/bin/x86_64-conda-linux-gnu-cc -Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -O2 -Wall -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /opt/conda/envs/scipy-dev/include -fPIC

creating /tmp/tmp952zmxc9/tmp
creating /tmp/tmp952zmxc9/tmp/tmp952zmxc9
compile options: '-I/opt/conda/envs/scipy-dev/include/python3.9 -c'
extra options: '-std=c++14 -pthread -fvisibility=hidden'
ccache: /tmp/tmp952zmxc9/main.c
cc1: warning: command line option '-std=c++14' is valid for C++/ObjC++ but not for C
building 'scipy.fft._pocketfft.pypocketfft' extension
compiling C++ sources
C compiler: ccache /opt/conda/envs/scipy-dev/bin/x86_64-conda-linux-gnu-cc -Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -O2 -Wall -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /opt/conda/envs/scipy-dev/include -fPIC

compile options: '-DPOCKETFFT_PTHREADS -I/opt/conda/envs/scipy-dev/lib/python3.9/site-packages/pybind11/include -I/opt/conda/envs/scipy-dev/lib/python3.9/site-packages/pybind11/include -I/opt/conda/envs/scipy-dev/lib/python3.9/site-packages/numpy/core/include -Ibuild/src.linux-x86_64-3.9/numpy/distutils/include -I/opt/conda/envs/scipy-dev/include/python3.9 -c'
extra options: '-std=c++14 -pthread -fvisibility=hidden -msse -msse2 -msse3'
ccache -pthread -shared -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath,/opt/conda/envs/scipy-dev/lib -Wl,-rpath-link,/opt/conda/envs/scipy-dev/lib -L/opt/conda/envs/scipy-dev/lib -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath,/opt/conda/envs/scipy-dev/lib -Wl,-rpath-link,/opt/conda/envs/scipy-dev/lib -L/opt/conda/envs/scipy-dev/lib -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath,/opt/conda/envs/scipy-dev/lib -Wl,-rpath-link,/opt/conda/envs/scipy-dev/lib -L/opt/conda/envs/scipy-dev/lib -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/envs/scipy-dev/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /opt/conda/envs/scipy-dev/include build/temp.linux-x86_64-3.9/scipy/fft/_pocketfft/pypocketfft.o -Lbuild/temp.linux-x86_64-3.9 -o build/lib.linux-x86_64-3.9/scipy/fft/_pocketfft/pypocketfft.cpython-39-x86_64-linux-gnu.so -Wl,--version-script=build/temp.linux-x86_64-3.9/link-version-scipy.fft._pocketfft.pypocketfft.map -pthread
ccache: invalid option -- 't'
Usage:
```

**Comment by trallard** [comment]
👀  on it!

**Comment by rgommers** [comment]
I'm not sure it's super urgent, people can run tests with `pytest scipy`. We can fix it after the sprint.

**Comment by trallard** [comment]
Will have a quick look - if it can be easily fixed I can send a PR otherwise will leave for later 😉

File: tools/docker_dev/gitpod.Dockerfile [hunk 1]
```diff
@@ -5,6 +5,8 @@ FROM ${BASE_CONTAINER} as clone
 
 COPY --chown=gitpod . /tmp/scipy_repo
 RUN git clone --depth 1 file:////tmp/scipy_repo /tmp/scipy
+WORKDIR /tmp/scipy
+RUN git submodule update --init --depth=1 -- scipy/_lib/boost
```
Comment (trallard) [inline tools/docker_dev/gitpod.Dockerfile:9]:
```suggestion
```

File: tools/docker_dev/gitpod.Dockerfile [hunk 2]
```diff
@@ -5,6 +5,8 @@ FROM ${BASE_CONTAINER} as clone
```
Comment (trallard) [inline tools/docker_dev/gitpod.Dockerfile:5]:
```suggestion
# note this only needs to be done for the scipy repo
# submodules can be dealt with in the runtime image
```

**Review by trallard — CHANGES_REQUESTED** [review]
Since I cannot push on top of `HEAD`

I'd move this to right after line 37  in `gitpod.Dockerfile` https://github.com/scipy/scipy/blob/d8379ea8689df5d4ce6e21c43d95d37974313f8a/tools/docker_dev/gitpod.Dockerfile#L36

```# Need to initialise the boost submodule
RUN git submodule update --init --depth=1 -- scipy/_lib/boost
```

And add right after line 16 in `.gitpod.yml` https://github.com/scipy/scipy/blob/28601b95720474a4f8701d6cb784ac6ec3618d49/.gitpod.yml#L15

```
# need to reinitialise the submodule due to how GitPod handles repos
git submodule update --init --depth=1 -- scipy/_lib/boost
```

With this it *should* work - of course Gitpod should be happy about it

- I added those changes
- Spun up the container locally
- re-init the submodule && built Scipy
- `python runtests.py` now works 🎉

**Comment by trallard** [comment]
Notice `python runtests.py` has to be done after the build - 🤷🏽  I tried instead of the build and failed (which should not) but my guess is also the `CFLAGs` set for `ccache`? 

But when I first added `ccache` without the flags in `workspace_config` Docker did not know what to make of it

**Comment by rgommers** [comment]
Thanks Tania! I made those changes and merged, let's see if that fixes it.

**Comment by trallard** [comment]
This fixed the build. I had limited time to check Gitpod itself but can make sure things are working today

**Pull request merged**